src/Hint/NewType.hs

```diff
@@ -26,39 +31,110 @@ newtype Foo = Foo Int deriving stock Show
+newTypeDerivingStrategiesHintDecl _ = []
 
-singleSimpleField :: Decl_ -> Maybe (DataOrNew S, Type_, DataOrNew S -> Type_ -> Decl_)
-singleSimpleField (DataDecl x1 dt x2 x3 [QualConDecl y1 Nothing Nothing ctor] x4)
-    | Just (t, ft) <- f ctor = Just (dt, t, \dt t -> DataDecl x1 dt x2 x3 [QualConDecl y1 Nothing Nothing $ ft t] x4)
-    where
-        f (ConDecl x1 x2 [t]) | not $ isKindHash t = Just (t, \t -> ConDecl x1 x2 [t])
-        f (RecDecl x1 x2 [FieldDecl y1 [y2] t]) = Just (t, \t -> RecDecl x1 x2 [FieldDecl y1 [y2] t])
-        f _ = Nothing
+hasAllStrategies :: HsDataDefn GhcPs -> Bool
+hasAllStrategies (HsDataDefn _ NewType _ _ _ _ (L _ xs)) = all hasStrategyClause xs
+hasAllStrategies _ = False
+
+isData :: HsDataDefn GhcPs -> Bool
+isData (HsDataDefn _ NewType _ _ _ _ _) = False
+isData (HsDataDefn _ DataType _ _ _ _ _) = True
+isData _ = False
+
+hasStrategyClause :: LHsDerivingClause GhcPs -> Bool
+hasStrategyClause (L _ (HsDerivingClause _ (Just _) _)) = True
+hasStrategyClause _ = False
+
+data WarnNewtype = WarnNewtype
+    { newDecl :: LHsDecl GhcPs
+    , insideType :: HsType GhcPs
+    }
+
+-- | Given a declaration, returns the suggested \"newtype\"ized declaration following these guidelines:
+-- * Types ending in a \"#\" are __ignored__, because they are usually unboxed primitives - @data X = X Int#@
+-- * @ExistentialQuantification@ stuff is __ignored__ - @data X = forall t. X t@
+-- * Constructors with (nonempty) constraints are __ignored__ - @data X a = (Eq a) => X a@
+-- * Single field constructors get newtyped - @data X = X Int@ -> @newtype X = X Int@
+-- * Single record field constructors get newtyped - @data X = X {getX :: Int}@ -> @newtype X = X {getX :: Int}@
+-- * All other declarations are ignored.
+singleSimpleField :: LHsDecl GhcPs -> Maybe WarnNewtype
+singleSimpleField (L loc (TyClD ext decl@(DataDecl _ _ _ _ dataDef@(HsDataDefn _ DataType _ _ _ [L _ constructor] _))))
+    | Just inType <- simpleCons constructor =
+        Just WarnNewtype
+              { newDecl = L loc $ TyClD ext decl {tcdDataDefn = dataDef
+                  { dd_ND = NewType
+                  , dd_cons = map (\(L consloc x) -> L consloc $ dropConsBang x) $ dd_cons dataDef
+                  }}
+              , insideType = inType
+              }
 singleSimpleField _ = Nothing
 
-newTypeDerivingStrategiesHintDecl :: Decl_ -> [Idea]
-newTypeDerivingStrategiesHintDecl x =
-    [rawIdeaN Ignore "Use DerivingStrategies" (srcInfoSpan $ ann x) (prettyPrint x) Nothing [] | lacksStrategy x]
+-- | Checks whether its argument is a \"simple constructor\" (see criteria in 'singleSimpleFieldNew')
+-- returning the type inside the constructor if it is. This is needed for bang/MagicHash analysis.
+simpleCons :: ConDecl GhcPs -> Maybe (HsType GhcPs)
+simpleCons (ConDeclH98 _ _ _ [] context (PrefixCon [L _ inType]) _)
+    | emptyOrNoContext context
```

Comment:
These guards are duplicated but I couldn't think of a good name to give them.

Reply:
In the absence of a good name, I'd have matched with a case and defined a local `where f = ...`